File: network/failures.py
"""
Failure and impairment models.
"""

from __future__ import annotations

from dataclasses import dataclass
import random

from network_model import NetworkGraph, Link, LinkProperties


@dataclass
class LinkImpairments:
    blackout_prob_per_s: float = 0.0  # probability a given link is down each second (Bernoulli per step)
    latency_spike_prob_per_s: float = 0.0
    latency_spike_add_s: float = 0.05
    loss_multiplier: float = 1.0


class FailureModel:
    """
    Stateless step-wise impairment application.
    For research-grade work later, you can make these stateful (Markov).
    """

    def __init__(self, impair: LinkImpairments):
        self.impair = impair

    def apply(self, graph: NetworkGraph, dt_s: float) -> None:
        p_black = 1.0 - (1.0 - self.impair.blackout_prob_per_s) ** float(max(0.0, dt_s))
        p_spike = 1.0 - (1.0 - self.impair.latency_spike_prob_per_s) ** float(max(0.0, dt_s))

        to_remove = []
        for u, v, link in list(graph.edges()):
            if self.impair.blackout_prob_per_s > 0 and random.random() < p_black:
                to_remove.append((u, v))
                continue

            add_lat = self.impair.latency_spike_add_s if (self.impair.latency_spike_prob_per_s > 0 and random.random() < p_spike) else 0.0
            loss = min(1.0, max(0.0, link.props.loss_prob * self.impair.loss_multiplier))
            props = LinkProperties(
                latency_s=float(link.props.latency_s + add_lat),
                bandwidth_mbps=float(link.props.bandwidth_mbps),
                loss_prob=float(loss),
            )
            graph.adjacency[u][v] = Link(a=u, b=v, distance_km=link.distance_km, props=props)
            graph.adjacency[v][u] = Link(a=v, b=u, distance_km=link.distance_km, props=props)

        for u, v in to_remove:
            if v in graph.adjacency.get(u, {}):
                del graph.adjacency[u][v]
            if u in graph.adjacency.get(v, {}):
                del graph.adjacency[v][u]

File: src/failures.c
/* failures.c — port of network/failures.py FailureModel.apply. */
#include "astra/failures.h"
#include <math.h>

void astra_failures_apply(NetworkGraph *g, const LinkImpairments *imp,
                          double dt_s, Rng *rng) {
    double dt = dt_s < 0.0 ? 0.0 : dt_s;
    double p_black = 1.0 - pow(1.0 - imp->blackout_prob_per_s, dt);
    double p_spike = 1.0 - pow(1.0 - imp->latency_spike_prob_per_s, dt);

    for (uint32_t e = 0; e < g->link_count; ++e) {
        Link *L = &g->links[e];
        if (!L->up) continue;

        if (imp->blackout_prob_per_s > 0.0 && rng_double(rng) < p_black) {
            L->up = 0;                                   /* blackout: drop edge */
            continue;
        }
        double add_lat = (imp->latency_spike_prob_per_s > 0.0 &&
                          rng_double(rng) < p_spike) ? imp->latency_spike_add_s : 0.0;
        double loss = L->props.loss_prob * imp->loss_multiplier;
        if (loss < 0.0) loss = 0.0; else if (loss > 1.0) loss = 1.0;
        L->props.latency_s += add_lat;
        L->props.loss_prob  = (float)loss;
    }
}
